File: ios/Plugin/Services/AppCheckService.swift
import Foundation
import Capacitor
import FirebaseCore
import FirebaseAppCheck

class AppCheckService {
    private weak var plugin: CAPPlugin?
    private var tokenRefreshListener: NSObjectProtocol?
    
    init(plugin: CAPPlugin) {
        self.plugin = plugin
    }
    
    func initialize(_ call: CAPPluginCall) {
        guard let provider = call.getString("provider") else {
            call.reject("Provider is required")
            return
        }
        
        let debugToken = call.getString("debugToken")
        let isTokenAutoRefreshEnabled = call.getBool("isTokenAutoRefreshEnabled", true)
        
        // Configure the provider factory
        switch provider {
        case "deviceCheck":
            #if !targetEnvironment(simulator)
            let providerFactory = DeviceCheckProviderFactory()
            AppCheck.setAppCheckProviderFactory(providerFactory)
            #else
            call.reject("DeviceCheck is not available on simulator")
            return
            #endif
            
        case "appAttest":
            if #available(iOS 14.0, *) {
                let providerFactory = AppAttestProviderFactory()
                AppCheck.setAppCheckProviderFactory(providerFactory)
            } else {
                call.reject("App Attest requires iOS 14.0 or later")
                return
            }
            
        case "debug":
            let providerFactory = DebugAppCheckProviderFactory()
            AppCheck.setAppCheckProviderFactory(providerFactory)
            
            if let debugToken = debugToken {
                UserDefaults.standard.set(debugToken, forKey: "FIRAppCheckDebugToken")
            }
            
        default:
            call.reject("Unsupported provider: \(provider)")
            return
        }
        
        // Get shared instance and configure
        let appCheck = AppCheck.appCheck()
        appCheck.isTokenAutoRefreshEnabled = isTokenAutoRefreshEnabled
        
        // Set up token refresh listener
        setupTokenRefreshListener()
        
        call.resolve()
    }
    
    func getToken(_ call: CAPPluginCall) {
        let forceRefresh = call.getBool("forceRefresh", false)
        
        AppCheck.appCheck().token(forcingRefresh: forceRefresh) { token, error in
            if let error = error {
                call.reject("Failed to get App Check token", nil, error)
                return
            }
            
            guard let token = token else {
                call.reject("No token received")
                return
            }
            
            call.resolve([
                "token": token.token,
                "expireTimeMillis": Int(token.expirationDate.timeIntervalSince1970 * 1000)
            ])
        }
    }
    
    func setTokenAutoRefreshEnabled(_ call: CAPPluginCall) {
        guard let enabled = call.getBool("enabled") else {
            call.reject("enabled parameter is required")
            return
        }
        
        AppCheck.appCheck().isTokenAutoRefreshEnabled = enabled
        call.resolve()
    }
    
    private func setupTokenRefreshListener() {
        tokenRefreshListener = NotificationCenter.default.addObserver(
            forName: .AppCheckAppCheckTokenDidChange,
            object: nil,
            queue: .main
        ) { [weak self] notification in
            if let token = notification.userInfo?[kFIRAppCheckTokenNotificationKey] as? AppCheckToken {
                self?.plugin?.notifyListeners("appCheckTokenChanged", data: [
                    "token": token.token,
                    "expireTimeMillis": Int(token.expirationDate.timeIntervalSince1970 * 1000)
                ])
            }
        }
    }
    
    deinit {
        if let listener = tokenRefreshListener {
            NotificationCenter.default.removeObserver(listener)
        }
    }
}

File: android/src/main/java/com/ahsanmahmood/capacitor/firebasekit/services/AppCheckService.kt
package com.ahsanmahmood.capacitor.firebasekit.services

import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.google.firebase.Firebase
import com.google.firebase.appcheck.AppCheckToken
import com.google.firebase.appcheck.FirebaseAppCheck
import com.google.firebase.appcheck.appCheck
import com.google.firebase.appcheck.playintegrity.PlayIntegrityAppCheckProviderFactory
import com.google.firebase.appcheck.safetynet.SafetyNetAppCheckProviderFactory
import com.google.firebase.appcheck.debug.DebugAppCheckProviderFactory
import com.google.firebase.initialize

class AppCheckService(private val plugin: Plugin) {
    
    private var appCheck: FirebaseAppCheck? = null
    private var tokenListener: FirebaseAppCheck.AppCheckListener? = null
    
    fun initialize(call: PluginCall) {
        val provider = call.getString("provider")
        val debugToken = call.getString("debugToken")
        val isTokenAutoRefreshEnabled = call.getBoolean("isTokenAutoRefreshEnabled", true) ?: true
        
        try {
            // Initialize Firebase if not already initialized
            if (Firebase.app == null) {
                Firebase.initialize(plugin.context)
            }
            
            appCheck = Firebase.appCheck
            
            when (provider) {
                "playIntegrity" -> {
                    appCheck?.installAppCheckProviderFactory(
                        PlayIntegrityAppCheckProviderFactory.getInstance()
                    )
                }
                "safetyNet" -> {
                    appCheck?.installAppCheckProviderFactory(
                        SafetyNetAppCheckProviderFactory.getInstance()
                    )
                }
                "debug" -> {
                    if (debugToken != null) {
                        System.setProperty("app.check.debug.token", debugToken)
                    }
                    appCheck?.installAppCheckProviderFactory(
                        DebugAppCheckProviderFactory.getInstance()
                    )
                }
                else -> {
                    call.reject("Invalid provider: $provider")
                    return
                }
            }
            
            appCheck?.setTokenAutoRefreshEnabled(isTokenAutoRefreshEnabled)
            
            call.resolve()
        } catch (e: Exception) {
            call.reject("Failed to initialize App Check", e)
        }
    }
    
    fun getToken(call: PluginCall) {
        val forceRefresh = call.getBoolean("forceRefresh", false) ?: false
        
        if (appCheck == null) {
            call.reject("App Check not initialized")
            return
        }
        
        val tokenTask = if (forceRefresh) {
            appCheck!!.getAppCheckToken(true)
        } else {
            appCheck!!.getAppCheckToken(false)
        }
        
        tokenTask
            .addOnSuccessListener { tokenResponse ->
                val result = JSObject()
                result.put("token", tokenResponse.token)
                result.put("expireTimeMillis", tokenResponse.expireTimeMillis)
                call.resolve(result)
            }
            .addOnFailureListener { e ->
                call.reject("Failed to get App Check token", e)
            }
    }
    
    fun setTokenAutoRefreshEnabled(call: PluginCall) {
        val enabled = call.getBoolean("enabled") ?: return
        
        if (appCheck == null) {
            call.reject("App Check not initialized")
            return
        }
        
        appCheck?.setTokenAutoRefreshEnabled(enabled)
        call.resolve()
    }
    
    fun addTokenChangedListener() {
        if (appCheck == null || tokenListener != null) return
        
        tokenListener = FirebaseAppCheck.AppCheckListener { token ->
            val event = JSObject()
            event.put("token", token.token)
            event.put("expireTimeMillis", token.expireTimeMillis)
            plugin.notifyListeners("appCheckTokenChanged", event)
        }
        
        appCheck?.addAppCheckListener(tokenListener!!)
    }
    
    fun removeTokenChangedListener() {
        tokenListener?.let {
            appCheck?.removeAppCheckListener(it)
            tokenListener = null
        }
    }
}